### PR TITLE
Share common code between Riot and Posix implementation of mainloop

### PR DESCRIPTION
**Pull request by zehortigoza**
Part of the code was being duplicated between those implementations
of mainloop.

Signed-off-by: José Roberto de Souza <jose.souza@intel.com>

### CODE REVIEW

File: src/lib/common/sol-mainloop-impl-common.c [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+int
+sol_mainloop_impl_timeout_compare(const void *data1, const void *data2)
+{
+    const struct sol_timeout_common *a = data1, *b = data2;
+
+    return sol_util_timespec_compare(&a->expire, &b->expire);
+}
+
+void
+sol_mainloop_impl_common_shutdown(void)
```
Comment (lpereira) [inline src/lib/common/sol-mainloop-impl-common.c]: I believe `_common_` infix is all that's needed for these functions: they're not platform-specific functions, so there's no need to use the `_impl_` prefix.

File: src/lib/common/sol-mainloop-impl-riot.c [hunk 2]
```diff
@@ -238,16 +87,16 @@ timex_set_until_next_timeout(timex_t *timex)
 void
 sol_mainloop_impl_run(void)
```
Comment (lpereira) [inline src/lib/common/sol-mainloop-impl-riot.c]: This should be the `sol_mainloop_run()` function, which should call `sol_mainloop_impl_iter()` (or other similarly-named function), left to be implemented by the platform-specific code. This way, there's no need to expose the `_common_` functions.

File: src/lib/common/sol-mainloop-impl-riot.c [hunk 3]
```diff
@@ -259,78 +108,31 @@ sol_mainloop_impl_run(void)
...
-    now = sol_util_timespec_get_current();
-    sol_util_timespec_sum(&now, &timeout->timeout, &timeout->expire);
-    ret = sol_ptr_vector_insert_sorted(&timeout_vector, timeout, timeout_compare);
-    SOL_INT_CHECK_GOTO(ret, != 0, clean);
-    return timeout;
-
-clean:
-    free(timeout);
-    return NULL;
+    return sol_mainloop_impl_common_timeout_add(timeout_ms, cb, data);
```
Comment (lpereira) [inline src/lib/common/sol-mainloop-impl-riot.c]: Is the function call here really needed? If all platforms are going to use the same functions for these, do we still need `_impl_` versions of these functions?

File: src/lib/common/sol-mainloop-impl-common.c (hunk 1 above)
Comment (mbelluzzo) [inline src/lib/common/sol-mainloop-impl-common.c]: +1 to that

File: src/lib/common/sol-mainloop-impl-riot.c (hunk 3 above)
Comment (ulissesf) [inline src/lib/common/sol-mainloop-impl-riot.c]: That's a good question. Right now POSIX just adds locking, and even that we could generalize to share more code. We won't support child watch on RIOT, I think, right?

File: src/lib/common/sol-mainloop-impl-riot.c (hunk 2 above)
Comment (ulissesf) [inline src/lib/common/sol-mainloop-impl-riot.c]: @lpereira, he's just following the abstraction already in place. We can however improve it. We can drop this and I take a look at pushing (maybe) just the threading part as implementation specific on each OS? This way we could share more code among POSIX and small OSes mainloops.

Comment (lpereira) [inline src/lib/common/sol-mainloop-impl-riot.c]: I think it might be a good idea to try that and see if it works, yes.

**Comment by zehortigoza** [comment]
Closing, we are going forward with https://github.com/solettaproject/soletta/pull/191

**Pull request closed without merging**